packages/task/src/browser/task-configurations.ts

```diff
@@ -0,0 +1,257 @@
+                                this.tasksMap.set(task.label, task);
+                            }
+                            resolve();
+                        })
+                        .catch(err => {
+                            this.logger.error(`Workspace: ${err}`);
+                        });
+                })
+                .catch(err => {
+                    this.logger.error(`Workspace: ${err}`);
+                });
+        });
+    }
+
+    protected readConfig(uri: string): Promise<TaskOptions[]> {
+        return new Promise((resolve, reject) => {
+            const tasksFound: TaskOptions[] = [];
+
+            // read tasks defined in tasks.json
+            this.readTasks(uri)
+                .then(tasks => {
+                    for (const task of tasks) {
+                        tasksFound.push(task);
+                    }
+                    resolve(tasksFound);
+                })
+                .catch(err => {
+                    reject(err);
+                });
+        });
+    }
+
+    protected getConfigFile(rootDir: string): Promise<string> {
+        return new Promise((resolve, reject) => {
+            // compute URL of the tasks file to watch
+            const cfgFile = new URI(rootDir).resolve(this.TASKFILEPATH).resolve(this.TASKFILE).toString();
+            this.fileSystem.exists(cfgFile).then(exists => {
+                if (exists) {
+                    resolve(cfgFile);
+                } else {
+                    this.logger.error(`config file tasks.json does not exist under ${rootDir}`);
+                    reject(`config file does not exist under: ${rootDir}`);
+                }
+            });
+        });
+    }
+
+    getTaskLabels(): string[] {
+        return [...this.tasksMap.keys()];
+    }
+
+    getTask(taskLabel: string): Promise<TaskOptions> {
+        return new Promise((resolve, reject) => {
+            if (this.tasksMap.has(taskLabel)) {
+                resolve(this.tasksMap.get(taskLabel));
+            } else {
+                reject(`Task with label ${taskLabel} not found`);
+            }
+        });
+    }
+
+    protected watchConfigFile(uri: string): Promise<number> {
+        // watch file for changes
+        return new Promise(resolve => {
+            this.watcherServer.watchFileChanges(uri)
+                .then(id => {
+                    this.toDispose.push(Disposable.create(() =>
+                        this.watcherServer.unwatchFileChanges(id))
+                    );
+                    resolve(id);
+                });
+        });
+    }
+
+    // a task config file we're watching has changed - update task list
+    protected onDidTaskFileChange(changes: DidFilesChangedParams): Promise<void> {
+        return new Promise((resolve, reject) => {
+            // one tasks.json file we're watching has changed?
+            if (this.hasWatchedFileChanged(changes.changes)) {
+                this.tasksMap.clear();
```

Comment:
It would be better to not trigger any changes when the underlying JSON is erroneous. Since we have auto-save (and it is on by default) keystrokes would clear the map unnecessary. The errors should be shown to the user through the editor, so it would not be surprising if the tasks have not been updated.

Reply:
good point. Done.